src/man/man3/Viewer.3
.TH Viewer 3U "6 October 1990" "Unidraw" "InterViews Reference Manual"
.SH NAME
Viewer \- displays a graphical component view hierarchy
.SH SYNOPSIS
.B #include <Unidraw/viewer.h>
.SH DESCRIPTION
A Viewer displays a graphical component view and provides an interface
to scrolling and zooming it.  The viewer also processes user input
events and generates the appropriate calls on Unidraw objects in
response.  The Viewer class is a derived from GraphicBlock; thus it
can draw the structured graphics hierarchy embodied in GraphicView's
graphic.  A viewer uses a damage object to keep the screen up to date.
Viewer's sensor catches up and down events, and its Handle implements
implements the tool manipulation loop.
.SH PUBLIC OPERATIONS
.TP
.B "Viewer("
.ns
.TP
.B "    Editor*, GraphicView*, Page*, Grid* = nil,"
.ns
.TP
.B "    Coord width = 0, Coord height = 0, Orientation = Normal,"
.ns
.TP
.B "    Alignment = Center, Zooming = Binary"
.ns
.TP
.B ")"
.ns
.TP
.B "Viewer("
.ns
.TP
.B "    const char*, Editor*, GraphicView*, Page*, Grid* = nil,"
.ns
.TP
.B "    Coord width = 0, Coord height = 0"
.ns
.TP
.B "    Alignment = Center, Zooming = Binary"
.ns
.TP
.B ")"
.br
Both constructors take the enclosing editor, the GraphicView object
that the viewer will display, a page object, and optionally a grid, a
natural width and height, initial page orientation, alignment, and
zooming specifications.  The second form adds a parameter for
specifying the Viewer's instance name.
.TP
.B "virtual void Update()"
Viewer augments GraphicBlock's Update semantics by repairing any
damage that has been accumulated.  In addition, Update will replace
its graphical view if its subject differs from the editor's component;
in that case, it will create a graphical view of the editor's
component, delete its own view, and replace it with the new one.
.TP
.B "virtual void Adjust(Perspective&)"
Viewer's Adjust operation will automatically update the editor's
MagnifVar (if it defines one) to reflect the current magnification.
.TP
.B "virtual void Handle(Event&)"
By default, Handle employs the editor's current tool on a LEFTMOUSE
down event.  On MIDDLEMOUSE or RIGHTMOUSE down events, it momentarily
engages the editor's MoveTool or SelectTool, respectively, allowing
the user to move or select a graphical component without changing the
current tool explicitly.  Nothing happens if the editor does not
define a keyboard equivalent for engaging these tools.  The Handle
function calls either the UseTool or MomentaryUseTool protected member
functions (described below) in response to these events.
.TP
.B "virtual void SetGraphicView(GraphicView*)"
.ns
.TP
.B "virtual void SetPage(Page*)"
.ns
.TP
.B "virtual void SetGrid(Grid*)"
.ns
.TP
.B "virtual GraphicView* GetGraphicView()"
.ns
.TP
.B "virtual Page* GetPage()"
.ns
.TP
.B "virtual Grid* GetGrid()"
.ns
.TP
.B "virtual Editor* GetEditor()"
Explicitly set or get a constructor-specified parameter.  The Set
operations delete the value they replace if it is different from the
given one.
.TP
.B "virtual void SetMagnification(float)"
Viewer's SetMagnification operation amends
GraphicBlock::SetMagnification to set the value of the enclosing
editor's MagnifVar, if it defines one.
.TP
.B "virtual void SetOrientation(Orientation)"
.ns
.TP
.B "virtual Orientation GetOrientation()"
Set and get the viewer's orientation parameter, which specifies
whether the graphical view, page, and grid will appear in portrait or
landscape mode.  No transformation will be applied in portrait mode,
while landscape mode will rotate the page \-90 degrees and translate it
vertically by an amount equal to the width of the page's bounding box.
.TP
.B "virtual Selection* GetSelection()"
Return the enclosing editor's selection object; it is equivalent to
calling GetEditor and then calling GetSelection on the result.
.TP
.B "virtual Damage* GetDamage()"
Return the viewer's damage object.
.TP
.B "virtual void InitRubberband(Rubberband*)"
.ns
.TP
.B "virtual void InitTextDisplay(TextDisplay*, Painter*)"
These operations let Rubberband and TextDisplay instances appear to
work in a viewer by allowing them to draw on the viewer's canvas.
InitRubberband sets the rubberband's canvas to the viewer's and
initializes its painter as well.  InitTextDisplay likewise sets the
TextDisplay's canvas to the viewer's.  It also modifies the given
painter's background color to match the viewer's and supplies the
painter to the TextDisplay for its use.
.TP
.B "virtual void IncurTextDisplayDamage(TextDisplay*, Painter*)"
Incur minimal damage on the viewer's damage object to account for the
TextDisplay's current size.  The TextDisplay's painter must be passed
explicitly.
.TP
.B "virtual void CenterOp()"
.ns
.TP
.B "virtual void Magnify(Coord, Coord, Coord, Coord)"
.ns
.TP
.B "virtual void ReduceToFit()"
Convenience functions for adjusting the viewer's perspective.
CenterOp centers the page in the viewer's canvas without changing the
magnification.  Magnify scrolls and magnifies the given area (in
canvas coordinates) so that it fills the viewer's canvas.  Magnify
ensures that the area will remain fully visible in the viewer.
ReduceToFit scrolls and reduces the view so that the entire page is
visible.
.TP
.B "virtual void Constrain(Coord&, Coord&)"
Constrain the supplied coordinates to fall on the nearest grid point
if the viewer defines a grid and if the enclosing editor defines an
active GravityVar.
.TP
.B "virtual void UseTool(Tool*)"
Commence direct-manipulation with the given tool without an explicit
action on the user's part.  Normally, direct manipulation with the
currently engaged tool (as returned by the editor's GetCurTool
operation) starts automatically when the user clicks in the viewer.
.TP
.B "virtual void Align(GraphicComp*, Alignment)"
Align the graphical component with respect to viewer's canvas,
assuming a view of the component appears in the viewer.  For example,
a ``Center'' alignment will align the component's center with the
canvas'; a ``BottomLeft'' alignment will align the component's
bottom-left corner to the canvas'.
.SH PROTECTED OPERATIONS
.TP
.B "void Init("
.ns
.TP
.B "    Editor*, GraphicView*, Page*, Grid*,"
.ns
.TP
.B "    Coord, Coord, Orientation"
.ns
.TP
.B ")"
.ns
.TP
.B "void Init(Editor*, GraphicView*, Page*, Grid*)"
.br
Initialize the viewer's member variables.  The first form of this
operation initializes all parameters that can be passed to a
constructor.  The first form calls the second form, which initializes
a subset of the attributes that may change independently of the
others.
.TP
.B "Tool* CurTool()"
A convenience function equivalent to calling the same operation on the
enclosing editor.
.TP
.B "Transformer* GetTransformer()"
Return the transformer that the viewer uses to implement scrolling and
zooming.
.TP
.B "virtual void Manipulate(Manipulator*, Event&)"
This operation implements a direct manipulation loop governed by the
given manipulator.  First it calls Grasp on the manipulator with the
given event.  Then it reads events in a loop, calling the
manipulator's Manipulating operation with each new event.  When
Manipulating returns false, it calls Effect on the manipulator with
the event read last.
.TP
.B "virtual void UseTool(Tool*, Event&)"
.ns
.TP
.B "virtual void MomentaryUseTool(Tool*, Event&)"
Helper functions for initiating direct manipulation with tools and
executing their effect.  UseTool first instructs the given tool to
create a manipulator, and then it calls Manipulate to carry the
manipulation through to completion.  Next it asks the tool to
interpret the manipulator, which in turn may produce a command.
Finally, UseTool executes and logs the command produced.
MomentaryUseTool simply engages the given tool, calls UseTool, and
re-engages the original tool.
.TP
.B "virtual void Reorient()"
A helper function that rotates the viewer's graphic to reflect its
current orientation.
.TP
.B "virtual void UpdateMagnifVar()"
A helper function used in SetMagnification to notify the enclosing
viewer's MagnifVar (if it defines one) of the change in magnification.
.TP
.B "virtual ClassId ViewCategory()"
Returns the view category (see classes(3U)) the viewer will use to
instantiate a view if it is not given one explicitly.  ViewCategory
returns COMPONENT_VIEW by default.
.SH SEE ALSO
Damage(3U), Editor(3U), Event(3I), Graphic(3U), GraphicBlock(3U),
GraphicComp(3U), GraphicView(3U), Grid(3U), Manipulator(3U),
MoveTool(3U), Page(3U), Perspective(3I), Rubband(3I), SelectTool(3U),
Selection(3U), Sensor(3I), TextDisplay(3I), Tool(3U), classes(3U),
globals(3U), statevars(3U)


CHANGES
May 12th 2012  ivtools-1.2.10

Many practical and small extensions to comterp, the scripting language embedded in
comdraw, flipbook, graphdraw, and drawserv.

Dropped complete support for gcc 3.0 and earlier. Things still might work but you'd be better off
upgradeing to gcc 4.0 or staying with ivtools-1.2.9.

Feb. 23rd, 2010  ivtools-1.2.9

- Debian related changes suggested by Agustin Martin.

- usec timeout on comdraw pause command -- allows for writing loops
  that can be exited.

- option for an end-of-line token introduced to LexScan.

- port to 64 bit Fedora

- fix problem with negative numbers in attribute editor

- make File/Save and File/Revert always work regardless of modified flag

Nov. 21st, 2009  ivtools-1.2.8

- significant work on keeping up with gcc and libstdc++.  A wrapper
  for stdio.h was inadvertently getting pulled in by /usr/include
  files, causing a lot of trouble.  Now builds with gcc-4.4 on Ubuntu
  9.10.

August 25th, 2009  ivtools-1.2.7

- series of memory leak checking commands: alistleak, attrvleak, commandleak, compleak, graphlicleak, mlineleak, viewleak.  Each returns the current number of allocated objects of that type (AttributeValueList, AttributeValue, Command, OverlayComp, Graphic, MultiLineObj, and OverlayView).

- added reference counting to OverlayView's (ComponentView's) referred from AttributeValue's (#define RESOURCE_COMPVIEW)

- fixed attrname func.

- fixed stddev func.

- added BooleanType to NumFunc promotion method.

- fixed graphdraw copy/paste of graph fragments

- new SocketObj for use comterp

- fix behavior of symbol manipulating funcs, by using internal bquote.

- port to Ubuntu Heron

- applied (most of) patch to build on Fedora 10 contributed by John Heidemann

- balance of changes to support ipl-1.0.4

July 9th, 2008  ivtools-1.2.6

- Debian contributed patch for iostreams backward compatibility

- bug fix for dispatcher.c contributed by Damon Permezel

- add mute command to comterp

- balance of changes to support ipl-1.0.2

March 4th, 2008  ivtools-1.2.5

- overall evolution to support use by ipl-1.0.1 (see
  http://sf.net/projects/ipl for further details)

July 20th, 2006  ivtools-1.2.4

- changes for compiling with gcc-4.0

- changes to socket-driven (ACE) use of comterp to support multi-line commands

- changes to allow re-use of scanner/parser with multiple operator tables

October 7th, 2005  ivtools-1.2.3

- add support for an external libtiff with a --with-tiff (and
-with-tiff-lib) configure arguments.

- changes necessary to compile with gcc-4.0

- changes necessary to compile with gcc-4.0 on top of ACE-5.4.7

- changes necessary to compile on suse-9.3 linux (amd64 platform).
Contributed by Paul Vixie.

- fixed bug in "addtool" command/mechanism in graphdraw and drawserv

- added support for color expressions to graphdraw (:colexpr), to
allow attribute-sensitive computation of a node's color.

January 25th, 2005  ivtools-1.2.2

If you submitted a patch in the past half year to ivtools that you
don't see reflected here, please resubmit.  I lost that e-mail.

- fix for problem writing to /tmp

- compilation fix on NetBSD for comterp.c

- compilation patch submitted by Michal Palczew.

July 13th, 2004  ivtools-1.2.1

Two crashes discovered and fixed by Michal Palczewski (mpalczew at
users.sourceforge.net).  One was a crash on exit after a delete with
idraw/drawtool, the other was a crash upon delete with graphdraw.
They turned out to be long-standing bugs in the order of object
deletion, problems that hadn't appeared with other versions of gcc.
RedHat must have adopted a more aggressive recycling of freed space in
a recent release.  Thanks Michal.


June 3rd, 2004  ivtools-1.2

ivtools-1.2 sets a new direction for the package, toward peer-to-peer
drawing chat.  ivtools' drawserv, the editor that incorporates every ivtool
mechanism to date (multi-frame, graph-editing, integrated scripting
language) has been extended to support a unique kind of
small-granularity shared whiteboard.  

This work was based on the pre-existing networked scripting capability
of ivtools' comdraw.  Then the pasting of any graphic was intercepted,
and the graphic serialized into a command that could be communicated
to other drawing editors to keep them in sync.

Then a connection management system was established, that supported
two-way peer-to-peer links between drawing editors.  In addition the
return path on each socket was used for detecting timeout, which
causes the connection to be dropped.

Finally, a distributed mutually-exclusive graphic selection mechanism
was created, where every drawing editor is aware of who currently has
a graphic selected or who had it selected last if it is not currently
selected.  When one drawing editor tries to select a graphic, it first
requests that permission from the last selector (after determining it
is unselected).

Now you can link to other drawing editors, draw graphics and see them
appear remotely (although only in black and white), and exercise the
distributed selection mechanism.  Four pieces remain to be developed:
1) support for full-up graphic state transmittal (it is sent, but not
interpreted on the remote end);, 2) support for transmission of the
complete set of drawing editor commands; 3) extending the distributed
mutual-exclusion mechanism to select pseudo-graphics to acquire global
permission for connection, pasting, moving front-to-back, and
grouping; and 4) a two-tier mechanism for invoking select(), that will
allow the bypass of selection requests while engaged in lengthy direct
manipulations.

If all this interests you, and you'd like to help, build this version
of ivtools with ACE support, subscribe to the ivtools-dev mailing, and
help me do some testing, programming, thinking, whatever.  Or visit
http://www.ivtools.org to make a donation to support this work.  Thanks.

Scott Johnston




src/Unidraw/viewer.c
/*
 * Copyright (c) 1990, 1991 Stanford University
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Stanford not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  Stanford makes no representations about
 * the suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * STANFORD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.
 * IN NO EVENT SHALL STANFORD BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * Viewer implementation.
 */

#include <Unidraw/classes.h>
#include <Unidraw/ctrlinfo.h>
#include <Unidraw/editor.h>
#include <Unidraw/globals.h>
#include <Unidraw/grid.h>
#include <Unidraw/iterator.h>
#include <Unidraw/keymap.h>
#include <Unidraw/kybd.h>
#include <Unidraw/manips.h>
#include <Unidraw/selection.h>
#include <Unidraw/statevars.h>
#include <Unidraw/uctrl.h>
#include <Unidraw/upage.h>
#include <Unidraw/viewer.h>

#include <Unidraw/Commands/transforms.h>

#include <Unidraw/Components/grcomp.h>
#include <Unidraw/Components/grview.h>

#include <Unidraw/Graphic/damage.h>
#include <Unidraw/Graphic/picture.h>

#include <Unidraw/Tools/tool.h>

#include <IV-2_6/InterViews/painter.h>
#include <IV-2_6/InterViews/perspective.h>
#include <IV-2_6/InterViews/rubband.h>
#include <IV-2_6/InterViews/sensor.h>
#include <IV-2_6/InterViews/shape.h>
#include <IV-2_6/InterViews/textdisplay.h>
#include <InterViews/transformer.h>

#include <IV-2_6/_enter.h>

#include <OS/math.h>

#include <string.h>

/*****************************************************************************/

static Painter* xorPainter;

/*****************************************************************************/

static boolean Different (Graphic* g1, Graphic* g2) {
    boolean different = true;

    if (
        g1->GetFgColor() == g2->GetFgColor() &&
        g1->GetBgColor() == g2->GetBgColor() &&
        g1->BgFilled() == g2->BgFilled() &&
        g1->GetPattern() == g2->GetPattern() &&
        g1->GetBrush() == g2->GetBrush() &&
        g1->GetFont() == g2->GetFont()
    ) {
        Transformer identity;
        Transformer* t1 = g1->GetTransformer();
        Transformer* t2 = g2->GetTransformer();

        if (t1 == t2) {
            different = false;
        } else if (t1 == nil) {
            different = *t2 != identity;
        } else if (t2 == nil) {
            different = *t1 != identity;
        } else {
            different = *t1 != *t2;
        }
    }
    return different;
}

/*****************************************************************************/

class ViewerGraphic : public Picture {
public:
    ViewerGraphic(GraphicView*);

    void Update();
private:
    void totalGSGraphic(Graphic*, Graphic&);    // should be in base class
private:
    GraphicView* _view;
};

ViewerGraphic::ViewerGraphic (GraphicView* view) {
    _view = view;
    Append(_view->GetGraphic());
}

void ViewerGraphic::Update () {
    GraphicComp* parent = (GraphicComp*) _view->GetGraphicComp()->GetParent();

    if (parent != nil) {
        FullGraphic gs;
        totalGSGraphic(parent->GetGraphic(), gs);

        if (Different(this, &gs)) {
            Damage* damage = _view->GetViewer()->GetDamage();
            
            damage->Incur(this);
            *(Graphic*)this = *(Graphic*)(&gs);
            damage->Incur(this);
        }
    }
}

void ViewerGraphic::totalGSGraphic (Graphic* g, Graphic& gs) {
    Graphic* parent = g->Parent();

    if (parent == nil) {
        concatGraphic(g, nil, g, &gs);

    } else {
        totalGSGraphic(parent, gs);
        concatGraphic(g, g, &gs, &gs);
    }
}

/*****************************************************************************/

ViewerView::ViewerView (GraphicView* g, UPage* page, Grid* grid, Viewer* v) {
    Picture* p = new Picture;

    if (grid != nil) {
        p->Append(grid->GetGraphic());
    }
    p->Append(page->GetGraphic());
    p->Append(_vg = new ViewerGraphic(g));

    SetGraphic(p);
    _vg->SetTag(this);

    _viewer = v;
    Update();
}

ViewerView::~ViewerView () {
    Graphic* g = GetGraphic();
    Iterator i;
    g->First(i);

    while (!g->Done(i)) {
        g->Remove(i);
    }
    _vg->First(i);
    _vg->Remove(i);
    delete _vg;
}

void ViewerView::Update () { _vg->Update(); }
Viewer* ViewerView::GetViewer () { return _viewer; }

/*****************************************************************************/

Viewer::Viewer (
    Editor* ed, GraphicView* gv, UPage* page, Grid* grid, 
    Coord w, Coord h, Orientation orientation,
    Alignment align, Zooming zoom
) : GraphicBlock(nil, 0, align, zoom) {
    Init(ed, gv, page, grid, w, h, orientation);
}

Viewer::Viewer (
    const char* instance,
    Editor* ed, GraphicView* gv, UPage* page, Grid* grid, 
    Coord w, Coord h, Orientation orientation,
    Alignment align, Zooming zoom
) : GraphicBlock(instance, nil, 0, align, zoom) {
    Init(ed, gv, page, grid, w, h, orientation);
}

void Viewer::Init(
    Editor* ed, GraphicView* gv, UPage* page, Grid* grid,
    Coord w, Coord h, Orientation orientation
) {    
    SetClassName("Viewer");
    if (xorPainter == nil) {
        xorPainter = new Painter;
        Ref(xorPainter);
    }
    _damage = new Damage;
    _orientation = orientation;
    input = new Sensor(updownEvents);

    Init(ed, gv, page, grid);
    shape->width = w;
    shape->height = h;
    SetCanvasType(CanvasSaveContents);
}

void Viewer::Init (Editor* ed, GraphicView* gview, UPage* page, Grid* grid) {
    _editor = ed;
    _gview = gview;
    _page = page;
    _grid = grid;

    _viewerView = new ViewerView(_gview, _page, _grid, this);
    _graphic = _viewerView->GetGraphic();
    _damage->SetGraphic(_graphic);

    Reorient();
    GraphicBlock::Init();
    UpdateMagnifVar();
}

Viewer::~Viewer () {
    delete _damage;
    delete _viewerView;
    delete _page;
    delete _grid;
    delete _gview;
}

void Viewer::Reconfig () { _damage->SetPainter(output); }

void Viewer::Update () {
    Selection* s = GetSelection();
    GraphicView* view = GetGraphicView();
    Component* viewComp = view->GetGraphicComp();
    Component* edComp = _editor->GetComponent();

    if (viewComp != edComp) {
        ComponentView* newView = edComp->Create(ViewCategory());

        if (newView->IsA(GRAPHIC_VIEW)) {
            edComp->Attach(newView);
            newView->Update();
            SetGraphicView((GraphicView*) newView);

        } else {
            delete newView;
        }

    } else {
	s->Hide(this);
        _viewerView->Update();
        GraphicBlock::UpdatePerspective();
	_damage->Repair();
        s->Show(this);
    }
}

void Viewer::Draw () {
    Selection* s = GetSelection();

    GraphicBlock::Draw();
    s->Init(this);
    s->Show(this);

    _damage->Reset();
}

void Viewer::Redraw (Coord l, Coord b, Coord r, Coord t) {
    GraphicBlock::Redraw(l, b, r, t);

    xorPainter->Clip(canvas, l, b, r, t);
    GetSelection()->Show(this);
    xorPainter->NoClip();
}

void Viewer::Resize () {
    GraphicBlock::Resize();
    _damage->SetCanvas(canvas);
    GetSelection()->Init(this);
}

void Viewer::GetGraphicBox (Coord& l, Coord& b, Coord& r, Coord& t) {
    _page->GetGraphic()->GetBox(l, b, r, t);
}

void Viewer::Adjust (Perspective& np) {
    GraphicBlock::Adjust(np);
    UpdateMagnifVar();
}    

void Viewer::Handle (Event& e) {
    Tool* tool = CurTool();

    if (tool != nil && e.eventType == DownEvent) {
	switch (e.button) {
        case LEFTMOUSE:
            UseTool(tool, e);
            break;

        case MIDDLEMOUSE:
            if (e.control) {
                GrabScroll(e);
            } else {
                MomentaryUseTool(CODE_MOVE, e);
            }
            break;

        case RIGHTMOUSE:
            if (e.control) {
                RateScroll(e);
            } else {
                MomentaryUseTool(CODE_SELECT, e);
            }
            break;
	}
    }
}

void Viewer::InitRubberband (Rubberband* r) {
    r->SetPainter(xorPainter);
    r->SetCanvas(canvas);
}

void Viewer::InitTextDisplay (TextDisplay* td, Painter* p) {
    if (!p->BgFilled()) {
        p->FillBg(true);
        p->SetColors(nil, output->GetBgColor());
    }
    td->Draw(p, canvas);
}

void Viewer::IncurTextDisplayDamage (TextDisplay* td, Painter* p) {
    Coord l, b, r, t;

    td->Bounds(l, b, r, t);
    --l; --b; ++r; ++t;

    Transformer* rel = p->GetTransformer();
    if (rel != nil) rel->TransformRect(l, b, r, t);

    _damage->Incur(l, b, r, t);
}

void Viewer::SetGraphicView (GraphicView* gv) {
    Perspective np = *perspective;

    GetEditor()->GetSelection()->Clear();
    delete _viewerView;
    delete _gview;

    _gview = gv;
    _viewerView = new ViewerView(_gview, _page, _grid, this);
    _graphic = _viewerView->GetGraphic();
    _damage->SetGraphic(_graphic);
    _damage->Incur(0, 0, 0, 0);                 // for detecting Draw in Adjust

    Reorient();
    GraphicBlock::Init();
    Adjust(np);

    if (_damage->Incurred()) {                  // Adjust didn't Draw
        Draw();
    }
}

void Viewer::SetPage (UPage* page) {
    if (_page != page) {
        delete _viewerView;
        delete _page;

        Init(_editor, _gview, page, _grid);
        GraphicBlock::Update();
    }
}

void Viewer::SetGrid (Grid* grid) {
    if (_grid != grid) {
        delete _viewerView;
        delete _grid;

        Init(_editor, _gview, _page, grid);
        GraphicBlock::Update();
    }
}

void Viewer::UpdateMagnifVar () {
    MagnifVar* magnifVar = (MagnifVar*) _editor->GetState("MagnifVar");
    if (magnifVar != nil) magnifVar->SetMagnif(GetMagnification());
}

void Viewer::SetMagnification (float m) {
    GraphicBlock::SetMagnification(m);
    UpdateMagnifVar();
}    

void Viewer::Reorient () {
    if (
        _orientation == Rotated ||
        _orientation == Landscape ||
        _orientation == Horizontal
    ) {
        Coord l, b, r, t;
        GetGraphicBox(l, b, r, t);
        Graphic* g = GraphicBlock::GetGraphic();

        g->Rotate(-90., l, b);
        g->Translate(0., r-l);
    }
}    

void Viewer::SetOrientation (Orientation o) {
    if (_orientation != o) {
        _orientation = o;

        Coord l, b, r, t;
        GetGraphicBox(l, b, r, t);
        Graphic* g = GraphicBlock::GetGraphic();

        if (
            _orientation == Normal ||
            _orientation == Portrait ||
            _orientation == Vertical
        ) {
            g->Rotate(90., l, b);
            g->Translate(t-b, 0.);

        } else if (
            _orientation == Rotated ||
            _orientation == Landscape ||
            _orientation == Horizontal
        ) {
            g->Rotate(-90., l, b);
            g->Translate(0., r-l);
        }
        GraphicBlock::Update();
    }
}

Graphic* Viewer::GetGraphic () { return _gview->GetGraphic(); }
GraphicView* Viewer::GetGraphicView () { return _gview; }
UPage* Viewer::GetPage () { return _page; }
Grid* Viewer::GetGrid () { return _grid; }
Orientation Viewer::GetOrientation () { return _orientation; }
Editor* Viewer::GetEditor () { return _editor; }
Selection* Viewer::GetSelection () { return _editor->GetSelection(); }
Damage* Viewer::GetDamage () { return _damage; }

void Viewer::CenterOp () {
    Perspective np = *perspective;
    np.curx = (np.width - np.curwidth)/2;
    np.cury = (np.height - np.curheight)/2;
    Adjust(np);
}

void Viewer::Magnify (Coord left, Coord bottom, Coord right, Coord top) {
  if( left != right || top != bottom) {
    Perspective np = *perspective;
    NormalRect(left, bottom, right, top);
    np.curx += left;
    np.cury += bottom;
    np.curwidth = Math::max(right - left, 1);
    np.curheight = Math::max(top - bottom, 1);
    Adjust(np);
  }
  else {
    Perspective np = *perspective;
    NormalRect(left, bottom, right, top);
    np.curx = np.curx + left - np.curwidth/4;
    np.cury = np.cury + bottom - np.curheight/4;
    np.curwidth = np.curwidth/2;
    np.curheight = np.curheight/2;
    Adjust(np);
  }
}

void Viewer::ReduceToFit () {
    Perspective np = *perspective;
    np.curx = np.x0;
    np.cury = np.y0;
    np.curwidth = np.width;
    np.curheight = np.height;
    Adjust(np);
}

void Viewer::Constrain (Coord& x, Coord& y) {
    if (_grid != nil) {
        GravityVar* grav = (GravityVar*) GetEditor()->GetState("GravityVar");

        if (grav != nil && grav->IsActive()) {
            _grid->Constrain(x, y);
        }
    }
}

Tool* Viewer::CurTool () { return _editor->GetCurTool(); }

void Viewer::Manipulate (Manipulator* m, Event& e) {
    Listen(allEvents);
    m->Grasp(e);

    /*
     * boolean b is just here to workaround a cfront 3.0 bug.
     */
    boolean b = false;
    do {
        Read(e);
	b = m->Manipulating(e);
    } while (b);

    m->Effect(e);
    Listen(input);
}

static Transformer* ComputeRel (Viewer* v, Transformer* t) {
    Transformer* rel = new Transformer;
    GraphicComp* comp = v->GetGraphicView()->GetGraphicComp();
    comp->GetGraphic()->TotalTransformation(*rel);
    rel->Postmultiply(t);
    return rel;
}    

Transformer* Viewer::ComputeGravityRel() {
  Transformer* rel = ComputeRel(this, _graphic->GetTransformer());
  if (_grid != nil) {
    GravityVar* grav = (GravityVar*) GetEditor()->GetState("GravityVar");
    
    if (grav != nil && grav->IsActive()) {

      /* if gravity is on, make sure all x,y points are mapped to */
      /* integers in the drawing coordinate space.  This makes sure */
      /* resultant graphic is on grid, regardless of what resolution */
      /* it was drawn at, after any possible pan/zoom combinations */
      /* This removes the error in the gravity mechanism introduced */
      /* by the rounding at the end of Grid::Constrain */

      float affine[6];
      rel->matrix(affine[0], affine[1], affine[2],
		  affine[3], affine[4], affine[5]);
      float xdelta = round(affine[4]) - affine[4];
      float ydelta = round(affine[5]) - affine[5];
      rel->Translate(xdelta, ydelta);

    }
  }
  return rel;
}

Transformer* Viewer::GetRel() {
    return ComputeRel(this, _graphic->GetTransformer());
}

void Viewer::UseTool (Tool* t) {
    Event e;
    e.target = nil;
    e.eventType = EnterEvent;
    UseTool(t, e);
}

void Viewer::UseTool (Tool* t, Event& e) {
    Transformer* relative = ComputeGravityRel();
    Manipulator* m = t->CreateManipulator(this, e, relative);

    if (m != nil) {
        Manipulate(m, e);
        Command* cmd = t->InterpretManipulator(m);

        if (cmd != nil) {
            cmd->Execute();

            if (cmd->Reversible()) {
                cmd->Log();
	    } else {
		delete cmd;
            }
        }
        delete m;
    }
    Unref(relative);
}

void Viewer::MomentaryUseTool (const char* keyCode, Event& e) {
    Tool* curTool= _editor->GetCurTool();
    const char* origCode = curTool->GetControlInfo()->GetKeyCode();
    KeyMap* keymap = _editor->GetKeyMap();
    
    if (strcmp(origCode, keyCode) == 0) {
	UseTool(CurTool(), e);		/* already using this tool */
    } else {
	keymap->Execute(keyCode);	/* engage tool momentarily */
	UseTool(CurTool(), e);
	keymap->Execute(origCode);	/* revert to original tool */
    }
}

Transformer* Viewer::GetTransformer () {
    Transformer* t = _graphic->GetTransformer();
    
    if (t == nil) {
	t = new Transformer;
	_graphic->SetTransformer(t);
        Unref(t);
    }
    return t;
}

float Viewer::LimitMagnification (float desired) {
    const float lo = 1./8.;
    const float hi = 16.;

    return (desired < lo) ? lo : (desired > hi) ? hi : desired;
}

ClassId Viewer::ViewCategory () { return COMPONENT_VIEW; }

void Viewer::Align (GraphicComp* comp, Alignment a) {
    Graphic* g = comp->GetGraphic();
    float cl, cb, cr, ct;
    g->GetBounds(cl, cb, cr, ct);

    Perspective* p = GetPerspective();
    float mag = GetMagnification();
    float vl = float(p->curx - p->x0) / mag;
    float vb = float(p->cury - p->y0) / mag;
    float vr = float(p->curx - p->x0 + p->curwidth - 1) / mag;
    float vt = float(p->cury - p->y0 + p->curheight - 1) / mag;

    float dx, dy;

    switch (a) {
    case TopLeft:
    case CenterLeft:
    case BottomLeft:
    case Left:
        dx = vl - cl;
        break;

    case TopCenter:
    case Center:
    case BottomCenter:
        dx = (vr + vl - cr - cl) / 2;
        break;

    case TopRight:
    case CenterRight:
    case BottomRight:
    case Right:
        dx = vr - cr;
        break;
    }
    
    switch (a) {
    case TopLeft:
    case TopCenter:
    case TopRight:
    case Top:
        dy = vt - ct;
        break;

    case CenterLeft:
    case Center:
    case CenterRight:
        dy = (vt + vb - ct - cb) / 2;
        break;

    case BottomLeft:
    case BottomCenter:
    case BottomRight:
    case Bottom:
        dy = vb - cb;
        break;
    }
    MoveCmd mvcmd(GetEditor(), dx, dy);
    comp->Interpret(&mvcmd);
}
